src/index.js

```diff
@@ -0,0 +1,748 @@
+   */
+  async uploadFileInChunks(file, fileId, size) {
+    const { body } = file;
+    // Developers can use this to track file upload progress
+    this._chunks = Math.ceil(size / FILE_CHUNK_SIZE);
+    this._chunkNumber = 0;
+
+    // Iterate over the chunks and send them
+    while (this._chunkNumber <= (this._chunks - 1)) {
+      const start = this._chunkNumber * FILE_CHUNK_SIZE;
+      const end = Math.min(start + FILE_CHUNK_SIZE, size);
+      const chunk = body.slice(start, end);
+      const sha256 = create256HexHash(chunk);
+
+      await this.api.send('POST', `v7/file_cmds/upload/${fileId}/chunk/${this._chunkNumber}`, chunk, {
+        additionalHeaders: {
+          'Content-SHA256': sha256
+        }
+      })
+        .catch(error => {
+          throw new Error(`[${error.message} - ${error.status}] Chunk ${this._chunkNumber} not uploaded: ${error.body}`, error);
+        });
+
+      this._chunkNumber++;
+    }
+
+    return this._chunks;
+  }
+
+  /**
+   * Uploads an arbitrarily sized buffer or stream file and returns the uploaded asset information
+   * @see {@link https://bynder.docs.apiary.io/#reference/upload-assets}
+   * @async
+   * @param {Object} file An object containing the id of the desired collection.
+   * @param {String} file.filename The file name of the file to be saved
+   * @param {Buffer|Readable} file.body The file to be uploaded. Can be either buffer or a read stream.
+   * @param {Number} file.length The length of the file to be uploaded
+   * @param {Object} file.data An object containing the assets' attributes
+   * @param {String} fileId Optional file UUID.V4 identifier
+   * @return {Promise<object>} The information of the uploaded file, including IDs and all final file urls.
+   */
+  async uploadFile(file) {
+    const { body, filename, data } = file;
+    const { brandId } = data;
+    const bodyType = bodyTypes.get(body);
+    const size = getLength(file);
+
+    if (!brandId) {
+      return rejectValidation('upload', 'brandId');
+    }
+
+    if (!filename) {
+      return rejectValidation('upload', 'filename');
+    }
+
+    if (!body || !bodyType) {
+      return rejectValidation('upload', 'body');
+    }
+
+    if (!size || typeof size !== 'number') {
+      return rejectValidation('upload', 'length');
+    }
+
+    this._chunks = undefined;
+    this._chunkNumber = undefined;
+    this._sha256 = create256HexHash(file.body);
+
+    try {
+      const fileId = await this.prepareUpload();
+      const chunks = await this.uploadFileInChunks(file, fileId, size);
+      const correlationId = await this.finaliseUpload(fileId, filename, chunks, size);
+      const asset = await this.saveAsset({...data, fileId});
+
+      this.sha256 = undefined;
+
+      return { fileId, correlationId, ...asset };
+    } catch (error) {
+      return Promise.reject({
+        status: error.status || 0,
+        message: error.message || `File not uploaded: ${filename}`
+      });
+    }
```

Comment:
Wow this file is pretty huge. Can't we move at least the upload-related stuff to a separate file?

Reply:
Actually, this file was heavily reduced in size mostly due to separation of concerns (the API client, constant, utilities, and others were moved to their own context).

Reply:
wow again :D Still would prefer to have it split up a bit more, but fine for now